In [244]:
import pandas as pd

# Define the path to the csv file
file_path = 'Cleaned-Vehicle-Crashes-NYC.csv'

# Use pandas to read the csv file
df = pd.read_csv(file_path)

# Display the first few rows of the dataframe
df.head()

/var/folders/t9/pnmnt1qj2430nn5t2hzkqjkc0000gn/T/ipykernel_88380/2949633630.py:7: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)


CRASH DATETIME    BOROUGH ZIP CODE   LATITUDE  LONGITUDE  \
0  2012-07-01 00:05:00  MANHATTAN    10036  40.762127 -73.997387   
1  2012-07-01 00:10:00   BROOKLYN    11223  40.588868 -73.972745   
2  2012-07-01 00:20:00   BROOKLYN    11215  40.677406 -73.983048   
3  2012-07-01 00:22:00      BRONX    10451  40.824188 -73.913840   
4  2012-07-01 00:23:00   BROOKLYN    11234  40.607179 -73.916414   

   NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   

    CONTRIBUTING FACTOR VEHICLE 1   CONTRIBUTING FACTOR VEHICLE 2  \
0  Driver Inattention/Distraction  Driver Inattention/Distraction   
1  Driver Inattention/Distraction                     Unspecified   
2         Prescription Medication                     Unspecified   
3   Failure to Yield Right-of-Way                Backing Unsafely   
4                     Unspecified                     Unspecified   

  VEHICLE TYPE CODE 1            VEHICLE TYPE CODE 2  
0   PASSENGER VEHICLE                            BUS  
1   PASSENGER VEHICLE  SPORT UTILITY / STATION WAGON  
2             UNKNOWN                        BICYCLE  
3   PASSENGER VEHICLE              PASSENGER VEHICLE  
4   PASSENGER VEHICLE              PASSENGER VEHICLE

In [ ]:
from sqlalchemy import create_engine
import pandas as pd

# Create a connection
engine = create_engine('postgresql://your_username:your_password@your_host/your_database')

# A sample query of all data from the "CrashData" table in the "VEHICLE CRASHES NYC" database
sql_query = "SELECT * FROM CrashData;"

# Execute the query and read it into a DataFrame
df = pd.read_sql_query(sql_query, con=engine)

print(df.head())


# Important Note

The data used in this notebook has been imported directly from a CSV file for the sake of simplicity and efficiency. However, to demonstrate our understanding and ability to work with SQL databases, and for the sake of the project requirement, we have included the SQL code right above that would normally import this data from a SQL database.

An SQL code that includes the creation of a SQL schema that matches the structure of our dataset has been created. This schema can be found in the "Data" folder under the name "Vehicle_Crashes_NYC_Schema.sql".

In [245]:
df['ZIP CODE'] = pd.to_numeric(df['ZIP CODE'], errors='coerce')

In [246]:
print(df.dtypes)

CRASH DATETIME                    object
BOROUGH                           object
ZIP CODE                         float64
LATITUDE                         float64
LONGITUDE                        float64
NUMBER OF PERSONS INJURED        float64
NUMBER OF PERSONS KILLED         float64
CONTRIBUTING FACTOR VEHICLE 1     object
CONTRIBUTING FACTOR VEHICLE 2     object
VEHICLE TYPE CODE 1               object
VEHICLE TYPE CODE 2               object
dtype: object


In [247]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

# List of columns to encode
cols_to_encode = ['BOROUGH', 'VEHICLE TYPE CODE 1', 'VEHICLE TYPE CODE 2', 'CONTRIBUTING FACTOR VEHICLE 1', 'CONTRIBUTING FACTOR VEHICLE 2']

for col in cols_to_encode:
    df[col] = le.fit_transform(df[col])

In [248]:
df.head(20)

CRASH DATETIME  BOROUGH  ZIP CODE   LATITUDE  LONGITUDE  \
0   2012-07-01 00:05:00        2   10036.0  40.762127 -73.997387   
1   2012-07-01 00:10:00        1   11223.0  40.588868 -73.972745   
2   2012-07-01 00:20:00        1   11215.0  40.677406 -73.983048   
3   2012-07-01 00:22:00        0   10451.0  40.824188 -73.913840   
4   2012-07-01 00:23:00        1   11234.0  40.607179 -73.916414   
5   2012-07-01 00:25:00        3   11375.0  40.732979 -73.844348   
6   2012-07-01 00:30:00        1   11233.0  40.670118 -73.922467   
7   2012-07-01 00:30:00        2   10007.0  40.714678 -74.005887   
8   2012-07-01 00:37:00        2   10017.0  40.755924 -73.974889   
9   2012-07-01 00:44:00        1   11220.0  40.641746 -74.013815   
10  2012-07-01 00:45:00        2   10022.0  40.758976 -73.964503   
11  2012-07-01 00:45:00        3   11416.0  40.681047 -73.849300   
12  2012-07-01 01:00:00        0   10456.0  40.828729 -73.914723   
13  2012-07-01 01:00:00        3   11434.0  40.679937 -73.776362   
14  2012-07-01 01:01:00        1   11217.0  40.681852 -73.980036   
15  2012-07-01 01:16:00        2   10014.0  40.732914 -74.003976   
16  2012-07-01 01:20:00        1   11208.0  40.677166 -73.877740   
17  2012-07-01 01:20:00        0   10470.0  40.895221 -73.880041   
18  2012-07-01 01:30:00        3   11102.0  40.767736 -73.918494   
19  2012-07-01 01:45:00        2   10018.0  40.753606 -73.987089   

    NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
0                         0.0                       0.0   
1                         0.0                       0.0   
2                         0.0                       0.0   
3                         0.0                       0.0   
4                         0.0                       0.0   
5                         0.0                       0.0   
6                         0.0                       0.0   
7                         1.0                       0.0   
8                         1.0                       0.0   
9                         1.0                       0.0   
10                        0.0                       0.0   
11                        0.0                       0.0   
12                        1.0                       0.0   
13                        0.0                       0.0   
14                        0.0                       0.0   
15                        0.0                       0.0   
16                        0.0                       0.0   
17                        0.0                       0.0   
18                        0.0                       0.0   
19                        0.0                       0.0   

    CONTRIBUTING FACTOR VEHICLE 1  CONTRIBUTING FACTOR VEHICLE 2  \
0                              11                             10   
1                              11                             55   
2                              42                             55   
3                              18                              5   
4                              56                             55   
5                              56                             55   
6                              56                             55   
7                              11                             55   
8                              56                             55   
9                              56                             55   
10                             42                             55   
11                             43                             55   
12                             41                             55   
13                             56                             55   
14                              6                             24   
15                             19                             55   
16                             56                             55   
17                             56                             55   
18                             56               

In [249]:
print('0 = Bronx, 1 = Brooklyn, 2 = Manhattan, 3 = Queens, 4 = Staten Island')

0 = Bronx, 1 = Brooklyn, 2 = Manhattan, 3 = Queens, 4 = Staten Island


In [250]:
# Convert the 'CRASH DATETIME' column to datetime
df['CRASH DATETIME'] = pd.to_datetime(df['CRASH DATETIME'])

# Create new features
df['HOUR'] = df['CRASH DATETIME'].dt.hour
df['DAY_OF_WEEK'] = df['CRASH DATETIME'].dt.dayofweek
df['MONTH'] = df['CRASH DATETIME'].dt.month
df['YEAR'] = df['CRASH DATETIME'].dt.year

df.head()

CRASH DATETIME  BOROUGH  ZIP CODE   LATITUDE  LONGITUDE  \
0 2012-07-01 00:05:00        2   10036.0  40.762127 -73.997387   
1 2012-07-01 00:10:00        1   11223.0  40.588868 -73.972745   
2 2012-07-01 00:20:00        1   11215.0  40.677406 -73.983048   
3 2012-07-01 00:22:00        0   10451.0  40.824188 -73.913840   
4 2012-07-01 00:23:00        1   11234.0  40.607179 -73.916414   

   NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   

   CONTRIBUTING FACTOR VEHICLE 1  CONTRIBUTING FACTOR VEHICLE 2  \
0                             11                             10   
1                             11                             55   
2                             42                             55   
3                             18                              5   
4                             56                             55   

   VEHICLE TYPE CODE 1  VEHICLE TYPE CODE 2  HOUR  DAY_OF_WEEK  MONTH  YEAR  
0                  457                   94     0            6      7  2012  
1                  457                  766     0            6      7  2012  
2                  676                   74     0            6      7  2012  
3                  457                  633     0            6      7  2012  
4                  457                  633     0            6      7  2012

In [251]:
df = df.drop('CRASH DATETIME', axis=1)

In [252]:
# Define the new order of the columns
cols = ['YEAR', 'MONTH', 'DAY_OF_WEEK', 'HOUR'] + [col for col in df.columns if col not in ['YEAR', 'MONTH', 'DAY_OF_WEEK', 'HOUR']]

# Reorder the columns
df = df[cols]

df.head()

YEAR  MONTH  DAY_OF_WEEK  HOUR  BOROUGH  ZIP CODE   LATITUDE  LONGITUDE  \
0  2012      7            6     0        2   10036.0  40.762127 -73.997387   
1  2012      7            6     0        1   11223.0  40.588868 -73.972745   
2  2012      7            6     0        1   11215.0  40.677406 -73.983048   
3  2012      7            6     0        0   10451.0  40.824188 -73.913840   
4  2012      7            6     0        1   11234.0  40.607179 -73.916414   

   NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   

   CONTRIBUTING FACTOR VEHICLE 1  CONTRIBUTING FACTOR VEHICLE 2  \
0                             11                             10   
1                             11                             55   
2                             42                             55   
3                             18                              5   
4                             56                             55   

   VEHICLE TYPE CODE 1  VEHICLE TYPE CODE 2  
0                  457                   94  
1                  457                  766  
2                  676                   74  
3                  457                  633  
4                  457                  633

In [253]:
print(df.dtypes)

YEAR                               int32
MONTH                              int32
DAY_OF_WEEK                        int32
HOUR                               int32
BOROUGH                            int64
ZIP CODE                         float64
LATITUDE                         float64
LONGITUDE                        float64
NUMBER OF PERSONS INJURED        float64
NUMBER OF PERSONS KILLED         float64
CONTRIBUTING FACTOR VEHICLE 1      int64
CONTRIBUTING FACTOR VEHICLE 2      int64
VEHICLE TYPE CODE 1                int64
VEHICLE TYPE CODE 2                int64
dtype: object


In [255]:
# Import necessary libraries
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

X = df.drop(['BOROUGH'], axis=1)
y = df['BOROUGH']

# Split data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize model
model = DecisionTreeClassifier(random_state=42)

# Train model
model.fit(X_train, y_train)

# Make predictions on the validation set
y_pred = model.predict(X_val)

# Evaluate model
accuracy = accuracy_score(y_val, y_pred)
print(f'Accuracy: {accuracy}')

Accuracy: 0.999563947592585


In [256]:
# Drop Columns
df = df.drop(['ZIP CODE', 'LATITUDE', 'LONGITUDE'], axis=1)

In [257]:
# Import necessary libraries
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

X = df.drop(['BOROUGH'], axis=1)
y = df['BOROUGH']

# Split data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize model
model = DecisionTreeClassifier(random_state=42)

# Train model
model.fit(X_train, y_train)

# Make predictions on the validation set
y_pred = model.predict(X_val)

# Evaluate model
accuracy = accuracy_score(y_val, y_pred)
print(f'Accuracy: {accuracy}')

Accuracy: 0.3104693140794224


In [269]:
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeClassifier

X = df.drop(['BOROUGH'], axis=1)
y = df['BOROUGH']

# Initialize the model
model = DecisionTreeClassifier(random_state=42)

# Initialize RFE
rfe = RFE(estimator=model, n_features_to_select=5, step=1)

# Fit RFE
rfe = rfe.fit(X, y)

# Get the features ranking
ranking = rfe.ranking_

# Create a dataframe to display feature and ranking
features_rank = pd.DataFrame({'Feature': X.columns, 'Ranking': ranking})

# Display the top features
print(features_rank.sort_values(by='Ranking'))

                         Feature  Ranking
0                           YEAR        1
1                          MONTH        1
2                    DAY_OF_WEEK        1
3                           HOUR        1
8            VEHICLE TYPE CODE 1        1
9            VEHICLE TYPE CODE 2        2
6  CONTRIBUTING FACTOR VEHICLE 1        3
7  CONTRIBUTING FACTOR VEHICLE 2        4
4      NUMBER OF PERSONS INJURED        5
5       NUMBER OF PERSONS KILLED        6


In [270]:
# Initialize the model
model = DecisionTreeClassifier(random_state=42)

# Train the model
model.fit(X, y)

# Get feature importances
importances = model.feature_importances_

# Create a DataFrame to display features and their importance scores
feature_importances = pd.DataFrame({'Feature': X.columns, 'Importance': importances})

# Display the feature importances
print(feature_importances.sort_values(by='Importance', ascending=False))

                         Feature  Importance
3                           HOUR    0.243796
1                          MONTH    0.222736
2                    DAY_OF_WEEK    0.174253
0                           YEAR    0.096699
8            VEHICLE TYPE CODE 1    0.085261
9            VEHICLE TYPE CODE 2    0.074995
6  CONTRIBUTING FACTOR VEHICLE 1    0.060218
7  CONTRIBUTING FACTOR VEHICLE 2    0.025371
4      NUMBER OF PERSONS INJURED    0.016365
5       NUMBER OF PERSONS KILLED    0.000306


In [272]:
# Import necessary libraries
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, BatchNormalization, Dropout
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping
import numpy as np

X = df.drop(['BOROUGH'], axis=1)
y = df['BOROUGH']

# Encode labels in column 'BOROUGH'.
labelencoder = LabelEncoder()
y = labelencoder.fit_transform(y)
y = to_categorical(y)

# Split data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Standardize the features to have zero mean and unit variance
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# Initialize model
model = Sequential()

# Add layers to model
model.add(Dense(256, input_dim=X_train.shape[1], activation='relu'))  # Increase the number of neurons
model.add(BatchNormalization())
model.add(Dropout(0.2))  # Add dropout layer
model.add(Dense(256, activation='relu'))  # Increase the number of neurons
model.add(BatchNormalization())
model.add(Dropout(0.2))  # Add dropout layer
model.add(Dense(y_train.shape[1], activation='softmax'))  # The output layer size should match the number of classes

# Compile model with a custom learning rate
opt = Adam(learning_rate=0.0005)  # Adjust the learning rate
model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])

# Define early stopping
early_stopping = EarlyStopping(monitor='val_loss', patience=3)

# Train model with a custom batch size and number of epochs
model.fit(X_train, y_train, epochs=50, batch_size=128, verbose=1, validation_data=(X_val, y_val), callbacks=[early_stopping])  # Increase the number of epochs and adjust the batch size

# Evaluate model
loss, accuracy = model.evaluate(X_val, y_val, verbose=0)
print(f'Accuracy: {accuracy}')

Epoch 1/50


/Users/davidskaff/anaconda3/lib/python3.11/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


6164/6164 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.3217 - loss: 1.5313 - val_accuracy: 0.3632 - val_loss: 1.4158
Epoch 2/50
6164/6164 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.3624 - loss: 1.4160 - val_accuracy: 0.3679 - val_loss: 1.3969
Epoch 3/50
6164/6164 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.3706 - loss: 1.4024 - val_accuracy: 0.3802 - val_loss: 1.3865
Epoch 4/50
6164/6164 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.3754 - loss: 1.3940 - val_accuracy: 0.3815 - val_loss: 1.3827
Epoch 5/50
6164/6164 ━━━━━━━━━━━━━━━━━━━━ 10s 2ms/step - accuracy: 0.3780 - loss: 1.3883 - val_accuracy: 0.3824 - val_loss: 1.3786
Epoch 6/50
6164/6164 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.3806 - loss: 1.3856 - val_accuracy: 0.3854 - val_loss: 1.3767
Epoch 7/50
6164/6164 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.3819 - loss: 1.3836 - val_accuracy: 0.3866 - val_loss: 1.3754
Epoch 8/50
6164/6164 ━━━━━━━━━━━━━━━━━━━━ 11s 2ms/step - accuracy: 0.3827 - loss: 1.3830 - val

In [275]:
# Import necessary libraries
from sklearn.metrics import classification_report
import numpy as np

# Predict classes using the model
y_pred = model.predict(X_val)
y_pred_classes = np.argmax(y_pred, axis=1)
y_true_classes = np.argmax(y_val, axis=1)

# Calculate precision, recall, f1-score, and accuracy
report = classification_report(y_true_classes, y_pred_classes)

print(report)

6164/6164 ━━━━━━━━━━━━━━━━━━━━ 2s 341us/step
              precision    recall  f1-score   support

           0       0.39      0.00      0.00     27308
           1       0.36      0.56      0.44     61195
           2       0.54      0.41      0.47     47650
           3       0.35      0.43      0.38     52959
           4       0.00      0.00      0.00      8112

    accuracy                           0.39    197224
   macro avg       0.33      0.28      0.26    197224
weighted avg       0.39      0.39      0.35    197224

6164/6164 ━━━━━━━━━━━━━━━━━━━━ 2s 330us/step
              precision    recall  f1-score   support

           0       0.39      0.00      0.00     27308
           1       0.36      0.56      0.44     61195
           2       0.54      0.41      0.47     47650
           3       0.35      0.43      0.38     52959
           4       0.00      0.00      0.00      8112

    accuracy                           0.39    197224
   macro avg       0.33      0.28      0.

# Model Development Process

We started with a **Decision Tree Classifier**. The model was trained on a dataset of New York City accidents. We split the data into training and validation sets, trained the model on the training set, and then evaluated its performance on the validation set. The initial model achieved an accuracy of **0.31** on the validation set.

To improve the model’s performance, we transitioned to a **Sequential model from Keras** and performed several rounds of hyperparameter tuning.

In the first round of tuning, we increased the complexity of the model by increasing the number of neurons in each layer from 64 to 128. We also increased the number of epochs from 10 to 20, and increased the batch size from 32 to 64. Additionally, we adjusted the learning rate to 0.001.

In the second round of tuning, we further increased the complexity of the model by increasing the number of neurons in each layer from 128 to 256. We also increased the number of epochs from 20 to 50, and increased the batch size from 64 to 128. We adjusted the learning rate to 0.0005, added dropout layers, and implemented early stopping to halt the training process when the model’s performance on the validation set stopped improving. After these steps, the model’s accuracy on the validation set increased to **0.4**.